### PR TITLE
Add Admin Web Interface to Edit DCDO Settings

### PR DESCRIPTION
**Pull request by KatieShipley**

### CODE REVIEW

**Comment by sureshc** [comment]
DCDO enables the usage of configuration settings that can be created, set, and modified at runtime (dynamically). It complements CDO, which provides static configuration settings that are initialized when the application starts. DCDO permits a dynamic configuration setting (key) to have several different Ruby data types: number, string, boolean, array, and hash. In production, DCDO uses DynamoDB for storage. While DynamoDB supports [many different formats](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html) for Item Attributes, DCDO stores all of its key values DynamoDB Attributes with the data type String. DCDO uses the Ruby gem `oj` to serialize/de-serialize its key values to JSON and stores the serialized JSON as a String in DynamoDB.

DCDO’s DynamoDB storage adapter uses `oj` to marshall/un-marshall key values:

https://github.com/code-dot-org/code-dot-org/blob/5af2e7f7286cde5c349d68ff30c1fe80f2b04103/lib/dynamic_config/adapters/dynamodb_adapter.rb#L24

https://github.com/code-dot-org/code-dot-org/blob/5af2e7f7286cde5c349d68ff30c1fe80f2b04103/lib/dynamic_config/adapters/dynamodb_adapter.rb#L41

Even on non-production environments, DCDO’s memory and file system adapters use `oj`:
https://github.com/code-dot-org/code-dot-org/blob/5af2e7f7286cde5c349d68ff30c1fe80f2b04103/lib/dynamic_config/adapters/memory_adapter.rb#L12

https://github.com/code-dot-org/code-dot-org/blob/5af2e7f7286cde5c349d68ff30c1fe80f2b04103/lib/dynamic_config/adapters/json_file_adapter.rb#L17

**Comment by sureshc** [comment]
I’m going to add some data type specific get/set tests here as part of this Pull Request:
https://github.com/code-dot-org/code-dot-org/blob/5af2e7f7286cde5c349d68ff30c1fe80f2b04103/lib/dynamic_config/adapters/json_file_adapter.rb

File: dashboard/app/controllers/dynamic_config_controller.rb [hunk 1]
```diff
@@ -65,4 +65,54 @@ def gatekeeper_set
...
+          when "Integer"
+            Integer(raw_value)
+          when "Float"
+            Float(raw_value)
+          when "Boolean"
+            raw_value.to_bool
+          when "String"
+            raw_value
+          else
+            new_value = JSON.parse(raw_value)
```
Comment (sureshc) [inline dashboard/app/controllers/dynamic_config_controller.rb]: Should we use `oj` to parse for consistency with how the DCDO storage adapters serialize and load values as they fetch/write? We’d have to figure out the equivalent `oj` Errors to rescue down below, though.

**Comment by cat5inthecradle** [comment]
Hey @KatieShipley,  @sureshc  mentioned that you're using `oj` because of some weird behavior in the `json` gem.

Looks like that might be improved with the new version that @Hamms is working on updating us to.

https://github.com/code-dot-org/code-dot-org/pull/46444

**Comment by Hamms** [comment]
Ahhh, I had been wondering why it was necessary to manually specify the type in this frontend. Yeah, with a working `json` gem it should be possible to dramatically simplify this

**Comment by cat5inthecradle** [comment]
reopening this, for review once we get some in-flight DCDO fixes out.

**Pull request closed without merging**